api/adaptive_load/adaptive_load.proto

```diff
@@ -0,0 +1,74 @@
+// Top-level session spec and output protos for the adaptive load controller.
+
+syntax = "proto3";
+
+package nighthawk.adaptive_load;
+
+import "api/adaptive_load/benchmark_result.proto";
+import "api/adaptive_load/config.proto";
+import "api/adaptive_load/metric_spec.proto";
+import "api/client/options.proto";
+import "google/protobuf/duration.proto";
+import "google/rpc/status.proto";
+import "validate/validate.proto";
+
+// Parameters describing the adjusting and testing stages of an adaptive load
+// session, which consists of a series of Nighthawk benchmarks probing for
+// the optimal load on the system, followed by a longer benchmark to validate
+// the values. Load adjustments are calculated by the selected StepController
+// plugin. Metrics can come from Nighthawk stats and counters or
+// platform-specific data sources via MetricsPlugins.
+message AdaptiveLoadSessionSpec {
+  // Settings for MetricsPlugins that obtain metrics from outside sources.
+  // An entry is required for every plugin referred to by metric_thresholds,
+  // other than the "builtin" plugin. Optional.
+  repeated MetricsPluginConfig metrics_plugin_configs = 1;
+  // Metrics and thresholds that determine load adjustments. The order of
+  // metrics is not significant. Required.
+  repeated MetricSpecWithThreshold metric_thresholds = 2 [(validate.rules).repeated .min_items = 1];
+  // Metrics that are collected and included in the output but not taken into
+  // account when adjusting the load. May be used for debugging or
+  // visualization. Optional.
+  repeated MetricSpec informational_metric_specs = 3;
+  // A proto describing Nighthawk Service traffic. See
+  // https://github.com/envoyproxy/nighthawk/blob/master/api/client/options.proto
+  //
+  // The adaptive load controller will return an error if the |duration| or
+  // |open_loop| fields are set within |nighthawk_traffic_options|. The
+  // controller will also be configured to overwrite at least one of the
+  // numerical fields during the search, such as requests_per_second, so any
+  // value of those fields specified here will be ignored.
+  //
+  // All other fields in |nighthawk_traffic_options| are passed through to the
+  // Nighthawk Service.
+  //
+  // Note that |concurrency| in |nighthawk_traffic_options| multiplies the total
+  // RPS sent.
+  //
+  // Required.
+  nighthawk.client.CommandLineOptions nighthawk_traffic_template = 4
```

Comment:
This seems the wrong level of abstraction. E.g. there are NH CLI options that don't relate to traffic shaping for example, in the future we want to add new options that aren't 1:1. I recommend a standalone proto for this, which is probably some strict subset of NH CLI to begin with.

Reply:
I see what you mean. CommandLineOptions is actually thought of as a misnamed version of this standalone proto [1]. But some CLI-specific info not relevant to Nighthawk Service has crept in: OutputFormat [2]. Nighthawk Service-specific info is also possible in the future.

My actual goal in this initial implementation is to collect exactly the information needed to call Nighthawk Service, since internally it needs to be able to do that. In the future, we might move away from Nighthawk Service but still need the underlying traffic spec.

@dubious90 has some plans to fix the entire issue. It may be a lower cost to do this PR with CommandLineOptions as-is and clean up everything at once. We also have some time pressure for adaptive load so we can't necessarily afford to do cleanups at the beginning.

[1] https://github.com/envoyproxy/nighthawk/blob/677b78359748a24b72ce492bcec1fde66bb61c7b/api/client/options.proto#L97

[2] https://github.com/envoyproxy/nighthawk/blob/677b78359748a24b72ce492bcec1fde66bb61c7b/api/client/options.proto#L127